proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,973 @@
+parameters, `env` (which can have one of the values `prod`, `canary`,
+or `test`) and `version` (which can have one of the values `v1`, `v2`,
+or `v3`).  Now let's say that there is a `RouteConfiguration` with one
+route that should be selected via the parameter `env=prod` and another
+route that should be selected via the parameter `version=v1`. Without
+this design, the server would need to actually provide the cross-product
+of these parameter values, so there will be 9 different variants of the
+resource, even though there are only 4 unique contents for the resource.
+However, this design instead allows the server to provide only the 4
+unique variants of the resource, with constraints allowing each client
+to get the appropriate one:
+
+<table>
+  <tr>
+    <th>Dynamic Parameter Constraints on Resource</th>
+    <th>Resource Contents</th>
+  </tr>
+
+  <tr>
+    <td>
+<code>{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod" invert:true},
+      {key:"version" value:"v1" invert:true}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}</code>
+    </td>
+    <td>
+      <ul>
+      <li>does <i>not</i> include the route for <code>env=prod</code>
+      <li>does <i>not</i> include the route for <code>version=v1</code>
+      </ul>
+    </td>
+  </tr>
+
+  <tr>
+    <td>
+<code>{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1" invert:true}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}</code>
+    </td>
+    <td>
+      <ul>
+      <li>does include the route for <code>env=prod</code>
+      <li>does <i>not</i> include the route for <code>version=v1</code>
+      </ul>
+    </td>
+  </tr>
+
+  <tr>
+    <td>
+<code>{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod" invert:true},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}</code>
+    </td>
+    <td>
+      <ul>
+      <li>does <i>not</i> include the route for <code>env=prod</code>
+      <li>does include the route for <code>version=v1</code>
+      </ul>
+    </td>
+  </tr>
+
+  <tr>
+    <td>
+<code>{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}</code>
+    </td>
+    <td>
+      <ul>
+      <li>does include the route for <code>env=prod</code>
+      <li>does include the route for <code>version=v1</code>
+      </ul>
+    </td>
+  </tr>
+
+</table>
+
+## Rationale
+
+This section documents limitations and design alternatives that we
+considered.
+
+### Limitation on Enhancing Matching in the Future
+
+One limitation of this design is that, because all xDS transport protocol
+implementations (clients, servers, and caching proxies) need to implement
+this matching behavior, it will be very difficult to add new matching
+behavior in the future.  Doing so will probably require some sort of
+client capability.
```

Comment:
I think we can add the client capabilities when the time comes.  It's not clear to me that we gain anything by doing it now, since they won't provide any benefit right now.